### DISCUSSION TITLE
How to make PHP script run first when opening a website?

### DISCUSSION THREAD

**Discussion by timoshechkin** [nuxt/nuxt#16444] [Questions]
How to make PHP script run first when opening a website? This is necessary in order to obtain the data of an authorized user before opening the site. This data is in the global variable $_SERVER['AUTH_USER'].

**Comment by TheAlexLichter** [comment]
You could use a [nitro middleware / server middleware](https://v3.nuxtjs.org/guide/directory-structure/server#server-middleware) to e.g. call your PHP script and get the info as JSON response to process it in your Nuxt app.

**Reply by timoshechkin** [reply]
How to correctly call a PHP script to pass the environment headers from the client to it (the most important thing is to get information about the authorized user, which is transmitted to the server and stored in the $_SERVER [ 'AUTH_USER' ] variable)?

**Reply by TheAlexLichter** [reply]
You can pass your headers e.g. [via `$fetch`](https://github.com/unjs/ohmyfetch#-adding-headers). As usual, without more specific examples I can only provide a rather general approach.

**Reply by timoshechkin** [reply]
Is it possible in NUXT or VUE to read the request headers on the server when accessing from the client, so that I would not use PHP at all?

**Reply by mittci** [reply]
https://www.jsdocs.io/package/h3#getHeader

in server/api/test.ts

```
/**
 * @file
 */

import { defineEventHandler } from 'h3'

export default defineEventHandler((event) => {
  const r = $fetch('your_api_url_page', { method: 'post', body: {secret_key: 'secret' } })
  console.log('response> ', r)

  const h1 = getHeader(event, 'header-name')
  const h2 = getHeaders(event)
  
  return h2
})

```

What you need?

**Reply by mittci** [reply]
< ?php 
    echo json_encode($_SERVER [ 'AUTH_USER' ])
 ? >

**Reply by mittci** [reply]
in vue component https://v3.nuxtjs.org/api/composables/use-request-headers

```
// Get all request headers
const headers = useRequestHeaders()
// Get only cookie request header
const headers = useRequestHeaders(['cookie'])
```

**Reply by timoshechkin** [reply]
> https://www.jsdocs.io/package/h3#getHeader
> 
> in server/api/test.ts
> 
> ```
> /**
>  * @file
>  */
> 
> import { defineEventHandler } from 'h3'
> 
> export default defineEventHandler((event) => {
>   const r = $fetch('your_api_url_page', { method: 'post', body: {secret_key: 'secret' } })
>   console.log('response> ', r)
> 
>   const h1 = getHeader(event, 'header-name')
>   const h2 = getHeaders(event)
>   
>   return h2
> })
> 
> ```
> 
> What you need?

I need to make automatic authorization on the site if the user is authorized in the OS.  When opening a PHP file with the code: "r_print($_SERVER);" in the PHP browser, I get all the global variables, including $_SERVER['REMOTE_USER'].  But when calling the script with useFetch I get a 401 error.  I think this is due to the fact that this request is not sent from the browser and the web server does not have data about the authorized user.
This variable is populated by the web server after the authorization check at the time of the request from the browser.
I don't know where to dig anymore...

**Reply by mittci** [reply]
for that exist Bearer Authentication, JWT.
if you use php session when you need to pass session cookie and other with $fetch


You also can create new api endpoint in php and fetch data from it (also need some access check, security)

$fetch('new_endpoint', { body: { userid: 1 }, method: post, headers: {'Authorization': 'Bearer <toke>'}  })

in php

$user = db_quert(select from users where id = $_POST[userid])

And using some methods to log in user

user_login($user)

print get_some_data_that_depend_on_auth_user()

**Comment by mittci** [comment]
Normally you need to use php to create rest api like https://petstore.swagger.io/  (click on Authorize button )
Or use unjs/nitro

**Reply by timoshechkin** [reply]
I'm trying to make pass-through authorization on the site for users authorized in the operating system.  To identify a user in the database, I need to get on the site his login under which he was authorized in the operating system.  This is the main problem!
In a simple project where the backend of the site is written in PHP, when a user opens a page (for example, index.php), I can get his login under which he is registered in the operating system (from the $_SERVER['REMOTE_USER'] variable).
How to get on the site the LOGIN under which the user logged in to his operating system?  The database on the site stores user data, including their logins, by which I need to identify them.

**Reply by timoshechkin** [reply]
Can make an entry point to the site in PHP where I get the remote_user, and then go to the nuxt application with the received value?
